### PR TITLE
add option to disable search and proxy

### PR DESCRIPTION
**Pull request by ikaven1024**
Signed-off-by: yingjinhui <yingjinhui@didiglobal.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In `karmada-search` component, both `search` and `proxy` will cache the sub cluster's resource in memory, it will take much physical resource. While not both features are used for users. So I add the options to enable/disable these.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
This shall be merged into release branches.

**Does this PR introduce a user-facing change?**:
```release-note
`karmada-search`: users now can use `--disable-search` and `--disable-proxy` options to disable search and proxy feature (default both are enabled).
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Seems you want a switch to control if `search`/`proxy` is enabled or not. 
I'm afraid the feature gate is inappropriate as a feature gate used to set up a switch that is mainly for experiment features, after the feature becomes mature, the relevant feature gate will be removed eventually. 

We can consider doing it by providing a command flag.

**Comment by ikaven1024** [comment]
@liys87x @XiShanYongYe-Chang @RainbowMango please take a look.

**Comment by RainbowMango** [comment]
/assign
note: @XiShanYongYe-Chang is on vocation and will get back next Monday.

**Comment by RainbowMango** [comment]
@ikaven1024 Please update the PR title, descriptions, and release note, since the current approach doesn't feature gate anymore.

**Comment by RainbowMango** [comment]
> In karmada-search component, both search and proxy will cache the sub cluster's resource in memory, it will take much physical resource. 

Is there a possibility that combining the two caches together?

**Comment by ikaven1024** [comment]
> Is there a possibility that combining the two caches together?

It's not easy to do this. @liys87x is studying it. https://github.com/karmada-io/karmada/issues/2432#issuecomment-1229853832.

Before this realized, we have to add options to disable them.

File: cmd/karmada-search/app/options/options.go [hunk 1]
```diff
@@ -69,6 +72,8 @@ func (o *Options) AddFlags(flags *pflag.FlagSet) {
 
 	flags.Float32Var(&o.KubeAPIQPS, "kube-api-qps", 40.0, "QPS to use while talking with karmada-apiserver. Doesn't cover events and node heartbeat apis which rate limiting is controlled by a different set of flags.")
 	flags.IntVar(&o.KubeAPIBurst, "kube-api-burst", 60, "Burst to use while talking with karmada-apiserver. Doesn't cover events and node heartbeat apis which rate limiting is controlled by a different set of flags.")
+	flags.BoolVar(&o.DisableSearch, "disable-search", o.DisableSearch, "turn off the search feature.")
+	flags.BoolVar(&o.DisableProxy, "disable-proxy", o.DisableProxy, "turn off the proxy feature.")
```
Comment (RainbowMango) [inline cmd/karmada-search/app/options/options.go:76]:
```suggestion
	flags.BoolVar(&o.DisableSearch, "disable-search", false, "Disable search feature that would save memory usage significantly.")
	flags.BoolVar(&o.DisableProxy, "disable-proxy", false, "Disable proxy feature that would save memory usage significantly.")
```
Just give my comments here, if we can't combine the two flags together for now.

By the way, what if both flags are `true`? What's the behavior?

Comment (ikaven1024) [reply cmd/karmada-search/app/options/options.go:76]: > By the way, what if both flags are `true`? What's the behavior?

Both search and proxy are working, they have two independent cache. For some large scale cluster, `karmada-search` may OOM.

File: pkg/search/controller.go [hunk 2]
```diff
@@ -115,7 +106,6 @@ func (c *Controller) Start(stopCh <-chan struct{}) {
 
 	defer runtime.HandleCrash()
 
-	c.informerFactory.Start(stopCh)
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:118]: Just a question: why do we need to delete this line?

Comment (ikaven1024) [reply pkg/search/controller.go:118]: karmadaInformerFactory is shared between searchController and proxyController. It's start is moved to
https://github.com/karmada-io/karmada/blob/b7554264ed07be3da20646d6e822f651415814dc/cmd/karmada-search/app/options/options.go#L108-L110

Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go:118]: ok, thanks for your explain~
/lgtm

**Comment by ikaven1024** [comment]
@RainbowMango shall we move forward it?

**Comment by RainbowMango** [comment]
Sure. I thought @XiShanYongYe-Chang has approved this.
/approve

**Pull request merged**